compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -633,46 +633,62 @@ impl<'a> AstValidator<'a> {
     /// - Non-const
     /// - Either foreign, or free and `unsafe extern "C"` semantically
     fn check_c_variadic_type(&self, fk: FnKind<'a>) {
-        let variadic_spans: Vec<_> = fk
-            .decl()
-            .inputs
-            .iter()
-            .filter(|arg| matches!(arg.ty.kind, TyKind::CVarArgs))
-            .map(|arg| arg.span)
-            .collect();
+        let variadic_params: Vec<_> =
+            fk.decl().inputs.iter().filter(|arg| matches!(arg.ty.kind, TyKind::CVarArgs)).collect();
 
-        if variadic_spans.is_empty() {
+        // The parser already rejects `...` if it's not the final argument, but we still want to
+        // emit the errors below, so we only consider the final `...` here.
+        let Some(variadic_param) = variadic_params.last() else {
             return;
+        };
+
+        if let Some(FnHeader { constness: Const::Yes(const_span), .. }) = fk.header() {
+            self.dcx().emit_err(errors::ConstAndCVariadic {
+                span: const_span.to(variadic_param.span),
+                const_span: *const_span,
+                variadic_span: variadic_param.span,
+            });
         }
 
-        if let Some(header) = fk.header() {
-            if let Const::Yes(const_span) = header.constness {
-                let mut spans = variadic_spans.clone();
-                spans.push(const_span);
-                self.dcx().emit_err(errors::ConstAndCVariadic {
-                    spans,
-                    const_span,
-                    variadic_spans: variadic_spans.clone(),
-                });
+        let (fn_ctxt, header) = match fk {
+            FnKind::Fn(fn_ctxt, _, Fn { sig, .. }) => (fn_ctxt, sig.header),
+            FnKind::Closure(..) => {
+                // Closures cannot be c-variadic (they can't even be `extern "C"`).
+                self.dcx().emit_err(errors::CVariadicClosure { span: variadic_param.span });
+                return;
```

Comment:
This is currently unreachable in practice, the parser already rejects `...` in closures. So, this could be a `span_bug` instead?

Reply:
huh, `span_bug` is only defined in `rustc_middle`, so we can't use that here. `panic!` or `unreachable!` also don't occur in this crate.

Reply:
The AST is weird on this point, yes.